libs/wingc/src/jsify.rs

```diff
@@ -135,26 +149,47 @@ impl<'a> JSifier<'a> {
 
 		if self.shim {
 			output.push(format!("const {} = require('{}');", STDLIB, STDLIB_MODULE));
-			output.push(format!("const $outdir = process.env.WING_SYNTH_DIR ?? \".\";"));
+			output.push(format!("const {} = process.env.WING_SYNTH_DIR ?? \".\";", OUTDIR_VAR));
+			output.push(format!("const {} = process.env.WING_TEST === \"true\";", ENV_WING_TEST));
 			output.push(TARGET_CODE.to_owned());
 		}
 
 		output.append(&mut imports);
 
 		if self.shim {
-			js.insert(
-				0,
-				format!(
-					"super({{ outdir: $outdir, name: \"{}\", plugins: $plugins }});\n",
-					self.app_name
-				),
-			);
-			output.push(format!(
-				"class MyApp extends {} {{\nconstructor() {}\n}}",
-				TARGET_APP,
-				JSifier::render_block(js)
+			let mut root_class = CodeMaker::default();
+			root_class.open(format!("class {} extends {} {{", ROOT_CLASS, STDLIB_CORE_RESOURCE));
+			root_class.open("constructor(scope, id) {");
+			root_class.line(format!("super(scope, id);"));
+			root_class.add_lines(js);
+			root_class.close("}");
+			root_class.close("}");
+
+			let mut app_wrapper = CodeMaker::default();
+			app_wrapper.open(format!("class {} extends {} {{", APP_CLASS, APP_BASE_CLASS));
+			app_wrapper.open("constructor() {");
+			app_wrapper.line(format!(
+				"super({{ outdir: {}, name: \"{}\", plugins: $plugins }});",
+				OUTDIR_VAR, self.app_name
 			));
-			output.push(format!("new MyApp().synth();"));
+			app_wrapper.open(format!("if ({}) {{", ENV_WING_TEST));
+			app_wrapper.line(format!("new {}(this, \"env0\");", ROOT_CLASS));
+			app_wrapper.line("const $test_runner = this.testRunner;".to_string());
```

Comment:
Perhaps there's a way to move more of this logic to the sdk instead of generate it per app?

Reply:
Sounds possible? For context, here's the high-level preflight code this PR generates at the moment:

```js
class $Root extends $stdlib.core.Resource {
	constructor(scope, id) {
		super(scope, id);
		// user resources
	}
}

class $App extends $AppBase { // $AppBase is one of the SDK's app classes
	constructor() {
		super({ outdir: $outdir, name: "test_bucket", plugins: $plugins });
		if ($wing_test) {
			new $Root(this, "env0");
			const $test_runner = this.testRunner;
			const $tests = $test_runner.findTests();
			for (let $i = 1; $i < $tests.length; $i++) {
				new $Root(this, "env" + $i);
			}
		} else {
			new $Root(this, "Default");
		}
	}
}
```

We could try something like this:

```js
class $App extends $AppBase { // $AppBase is one of the SDK's app classes
	constructor() {
		super({ outdir: $outdir, name: "test_bucket", plugins: $plugins, root: $Root, isTest: $wing_test });
	}
}
```

But I wasn't sure it made sense to make all of those pieces part of the `App`'s responsibilities. Definitely something we can play around with in the future though

Reply:
There's probably some clever-er solution I just can't think of it 😅

Reply:
I think passing the `$Root` type is legitimate (albeit not JSIIable).

Reply:
I agree with the point about testability, but I also still don't feel great about this `root` field in the `App` class / it doesn't seem right to me. :/ I also feel like some of this design might be clearer once we make some of the other modeling changes so that functions aren't materialized into every environment etc.

Reply:
No problem. Just some thoughts...